cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringServerHandler.java

```diff
@@ -139,6 +140,53 @@ public void connectControlChannel(FullHttpRequest request, HttpResponder respond
                                    TetheringControlResponse[].class));
   }
 
+  /**
+   * Sends control commands to the client and receives program status updates from the client.
+   */
+  @POST
+  @Path("/tethering/poll/{peer}")
+  public void pollControlChannel(FullHttpRequest request, HttpResponder responder, @PathParam("peer") String peer)
+    throws IOException, NotImplementedException, BadRequestException {
+    checkTetheringServerEnabled();
+    PeerInfo peerInfo;
+    try {
+      peerInfo = store.getPeer(peer);
+    } catch (PeerNotFoundException e) {
+      responder.sendString(HttpResponseStatus.NOT_FOUND, GSON.toJson(new TetheringErrorResponse(e.getMessage())));
+      return;
+    }
+    TetheringStatus tetheringStatus = peerInfo.getTetheringStatus();
+    store.updatePeerTimestamp(peer);
+    String messageId = processRequestContent(request);
+    List<TetheringControlMessage> controlMessages = new ArrayList<>();
+    MessageFetcher fetcher = messagingContext.getMessageFetcher();
+    TopicId topic = new TopicId(NamespaceId.SYSTEM.getNamespace(), topicPrefix + peer);
+    createTopicIfNeeded(topic);
+    String lastMessageId = messageId;
+    int batchSize = cConf.getInt(Constants.Tethering.CONTROL_MESSAGE_BATCH_SIZE);
+    try (CloseableIterator<Message> iterator =
+           fetcher.fetch(topic.getNamespace(), topic.getTopic(), batchSize, messageId)) {
+      while (iterator.hasNext()) {
+        Message message = iterator.next();
+        TetheringControlMessage controlMessage = GSON.fromJson(message.getPayloadAsString(StandardCharsets.UTF_8),
+                                                               TetheringControlMessage.class);
+        lastMessageId = message.getId();
+        controlMessages.add(controlMessage);
+      }
+    } catch (TopicNotFoundException e) {
+      String errorMsg = String.format("Peer %s not found", peer);
```

Comment:
we should be trying to create the topic over here, then pulling the control messages again (extract that part into a separate method so we don't duplicate code). If for some weird reason *another* TopicNotFoundException is thrown when pulling the control messages after we've created the topic here, then we can just return a 500 response code.

also add a comment about why we need to create the topic here.

Reply:
for future reference, it is cleaner if Handler classes basically contain no business logic. Normally they should only be validating the request, decoding things into java objects, calling some TetheringService class that has all the logic in it, then translating exceptions/responses into http responses.

This makes unit testing much simpler, as you can drop the http layer and just call methods directly, and also makes the TetheringService more modular and re-usable.

Reply:
fixed